File: Documentation/4.0/Samples/nodejs/indexes/querying/suggestions.js
import { 
    DocumentStore, 
    AbstractIndexCreationTask,
    MoreLikeThisStopWords,
    QueryData
} from "ravendb";

const store = new DocumentStore();
const session = store.openSession();

{

    //region suggestions_1
    class Products_ByName extends AbstractIndexCreationTask {
        constructor() {
            super();

            this.map = `from product in docs.Products select new { 
                product.Name 
            }`;

            this.index("Name", "Search"); // (optional) splitting name into multiple tokens
            this.suggestion("Name"); // configuring suggestions
        }
    }
    //endregion

    class Product { }

    async function suggestions() {
        
            {
                //region suggestions_2
                const product = await session
                    .query({ indexName: "Products/ByName" })
                    .search("Name", "chaig")
                    .firstOrNull();
                //endregion
            }

            {
                //region suggestions_3
                const suggestionResult = await session
                    .query({ indexName: "Products/ByName" })
                    .suggestUsing(builder => builder.byField("name", "chaig"))
                    .execute();

                console.log("Did you mean?");

                for (const suggestion of suggestionResult["name"].suggestions) {
                    console.log("\t" + suggestion);
                }
                //endregion
            }
 

        
            {
                //region query_suggestion_over_multiple_words
                const resultsByMultipleWords = await session
                    .query({ indexName: "Products/ByName" })
                    .suggestUsing(builder =>
                        builder.byField("name", [ "chaig", "tof" ])
                               .withOptions({
                                    accuracy: 0.4,
                                    pageSize: 5,
                                    distance: "JaroWinkler",
                                    sortMode: "Popularity"
                               }))
                    .execute();

                console.log("Did you mean?");

                for (const suggestion of resultsByMultipleWords["name"].suggestions) {
                    console.log("\t" + suggestion);
                }
                //endregion
            }
 
    }
}

File: Documentation/5.4/Samples/java/src/test/java/net/ravendb/Indexes/Querying/Suggestions.java
package net.ravendb.Indexes.Querying;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;
import net.ravendb.client.documents.indexes.FieldIndexing;
import net.ravendb.client.documents.queries.suggestions.StringDistanceTypes;
import net.ravendb.client.documents.queries.suggestions.SuggestionOptions;
import net.ravendb.client.documents.queries.suggestions.SuggestionResult;
import net.ravendb.client.documents.queries.suggestions.SuggestionSortMode;
import net.ravendb.client.documents.session.IDocumentSession;

import java.util.Map;

public class Suggestions {

    //region suggestions_1
    public class Products_ByName extends AbstractIndexCreationTask {
        public Products_ByName() {
            map = "from product in docs.Products " +
                "select new " +
                "{ " +
                "  product.Name " +
                "} ";

            index("Name", FieldIndexing.SEARCH); // (optional) splitting name into multiple tokens
            suggestion("Name");// configuring suggestions
        }
    }
    //endregion

    private class Product {

    }

    public Suggestions() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region suggestions_2
                Product product = session
                    .query(Product.class, Products_ByName.class)
                    .search("Name", "chaig")
                    .firstOrDefault();
                //endregion
            }

            try (IDocumentSession session = store.openSession()) {
                //region suggestions_3
                Map<String, SuggestionResult> suggestionResult = session
                    .query(Product.class, Products_ByName.class)
                    .suggestUsing(builder -> builder.byField("Name", "chaig"))
                    .execute();

                System.out.println("Did you mean?");

                for (String suggestion : suggestionResult.get("Name").getSuggestions()) {
                    System.out.println("\t" + suggestion);
                }
                //endregion
            }
        }

        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region query_suggestion_over_multiple_words
                SuggestionOptions options = new SuggestionOptions();
                options.setAccuracy(0.4f);
                options.setPageSize(5);
                options.setDistance(StringDistanceTypes.JARO_WINKLER);
                options.setSortMode(SuggestionSortMode.POPULARITY);

                Map<String, SuggestionResult> resultsByMultipleWords = session
                    .query(Product.class, Products_ByName.class)
                    .suggestUsing(builder ->
                        builder.byField("Name", new String[]{"chaig", "tof"})
                            .withOptions(options))
                    .execute();

                System.out.println("Did you mean?");

                for (String suggestion : resultsByMultipleWords.get("Name").getSuggestions()) {
                    System.out.println("\t" + suggestion);
                }
                //endregion
            }
        }
    }
}
